File: lib/snake/Segment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Direction_1 = require("../Direction");
const Point_1 = require("../Point");
const InvalidSegment_1 = require("./InvalidSegment");
class Segment {
    constructor(start, end) {
        this.start = start;
        this.end = end;
        if (this.start.x !== this.end.x && this.start.y !== this.end.y) {
            throw new InvalidSegment_1.default('no diagonals allowed');
        }
    }
    length() {
        return this.start.distance(this.end);
    }
    isEmpty() {
        return this.length() === 0;
    }
    direction(plane) {
        const point = this.norm();
        if (point !== undefined) {
            return Direction_1.fromPoint(point, plane);
        }
        else {
            return undefined;
        }
    }
    norm() {
        const x_length = this.end.x - this.start.x;
        const y_length = this.end.y - this.start.y;
        if (x_length) {
            return new Point_1.default(Math.abs(x_length) / x_length, 0);
        }
        else if (y_length) {
            return new Point_1.default(0, Math.abs(y_length) / y_length);
        }
        else {
            return undefined;
        }
    }
    connectsTo(other) {
        return this.end.equals(other.start);
    }
    equals(other) {
        return this.start.equals(other.start)
            && this.end.equals(other.end);
    }
    *points() {
        const norm = this.norm();
        if (norm !== undefined && norm.distance(new Point_1.default(0, 0)) === 1) {
            let cur = this.start;
            while (cur.equals(this.end) === false) {
                yield cur;
                cur = cur.add(norm);
            }
            yield cur; // last point
        }
    }
}
exports.default = Segment;

File: src/snake/Segment.py
from ..Point import Point, ZERO
from ..Direction import Direction, fromPoint

class InvalidSegment(Exception):
  pass

class Segment(object):
  def __init__(self, start, end):
    self.start = start
    self.end = end

    if self.start.x != self.end.x and self.start.y != self.end.y:
      raise InvalidSegment('no diagonals allowed')

  def length(self):
    return self.start.distance(self.end)

  def isEmpty(self):
    return self.length() == 0.0

  def norm(self):
    x_length = self.end.x - self.start.x
    y_length = self.end.y - self.start.y

    if x_length != 0:
      return Point(abs(x_length) / x_length, 0)
    elif y_length != 0:
      return Point(0, abs(y_length) / y_length)
    else:
      return None

  def direction(self, plane):
    point = self.norm()

    if point is not None:
      return fromPoint(point, plane)
    else:
      return None

  def connectsTo(self, other):
    return self.end == other.start

  def __eq__(self, other):
    if type(self) is type(other):
      return self.start == other.start and self.end ==  other.end
    else:
      return False

  def points(self):
    norm = self.norm()

    if norm is not None and norm.distance(ZERO) == 1:
      cur = self.start

      while cur != self.end:
        yield cur
        cur = cur.add(norm)

      yield cur; # last point

  def __str__(self):
    return '{}--{}'.format(str(self.start), str(self.end))

  def __repr__(self):
    return str(self)